vortex-array/src/builders/list.rs

```diff
@@ -0,0 +1,176 @@
+use std::any::Any;
+use std::sync::Arc;
+
+use vortex_dtype::{DType, Nullability, PType};
+use vortex_error::{VortexExpect, VortexResult};
+use vortex_scalar::{ListScalar, Scalar};
+
+use crate::array::ListArray;
+use crate::builders::{builder_with_capacity, ArrayBuilder, ArrayBuilderExt, BoolBuilder};
+use crate::validity::Validity;
+use crate::{ArrayData, IntoArrayData};
+
+pub struct ListBuilder {
+    value_builder: Box<dyn ArrayBuilder>,
+    index_builder: Box<dyn ArrayBuilder>,
+    validity: BoolBuilder,
+    nullability: Nullability,
+    dtype: DType,
+}
+
+impl ListBuilder {
+    pub fn with_capacity(
+        value_dtype: Arc<DType>,
+        nullability: Nullability,
+        capacity: usize,
+    ) -> Self {
+        let value_builder = builder_with_capacity(value_dtype.as_ref(), capacity);
+        let mut index_builder = if capacity < 2usize.pow(31) - 1 {
+            builder_with_capacity(
+                &DType::Primitive(PType::I32, Nullability::NonNullable),
+                capacity,
+            )
+        } else {
+            builder_with_capacity(
+                &DType::Primitive(PType::I64, Nullability::NonNullable),
+                capacity,
+            )
+        };
+
+        // The first index of the list, which is always 0 and represents an empty list.
+        index_builder.append_zero();
+
+        Self {
+            value_builder,
+            index_builder,
+            validity: BoolBuilder::with_capacity(Nullability::NonNullable, capacity),
+            nullability,
+            dtype: DType::List(value_dtype, nullability),
+        }
+    }
+
+    pub fn append_value(&mut self, value: ListScalar) -> VortexResult<()> {
+        if value.is_null() {
+            self.append_null();
+            Ok(())
+        } else {
+            for scalar in value.elements() {
+                // TODO(joe): This is slow, we should be able to append multiple values at once,
+                // or the list scalar should hold an ArrayData
+                self.value_builder.append_scalar(&scalar)?;
```

Comment:
is there a type check in the underlying builder?

Reply:
Yep, [see](https://github.com/spiraldb/vortex/blob/22ba94c47f9f98a939fb46cc9547d8700042c417/vortex-array/src/builders/mod.rs#L85)